src/index.js

```diff
@@ -1 +1,34 @@
-export default () => true
+import defaultsDeep from 'lodash.defaultsdeep';
+import pickBy from 'lodash.pickby';
+import makeOkCheck from './make-ok-check';
+import buildOptions from './build-options';
+import resolveUrl from './resolve-url';
+
+class Stipulate {
+
+  constructor(options) {
+    this.baseOptions = options;
+  }
+
+  beforeRequest(url, options) {
+    return [url, options];
+  }
+
+  afterResponse(response) {
+    return response;
+  }
+
+  send(urlString, options, query) {
+    const config = buildOptions(options, this.baseOptions);
+    const url = resolveUrl(urlString, query);
+
+    const requestArguments = this.beforeRequest(url, config);
+    const checkOk = makeOkCheck(config);
+
+    return fetch(...requestArguments)
+      .then(checkOk)
+      .then(this.afterResponse);
```

Comment:
>Then these aren't hooks. They're class methods intended to be overwritten.

Ok, I'll stop calling them hooks then. I thought it was pretty clear what I meant by using that word, but it sounds like you've got a conflicting definition in your mind, so we can drop the word entirely. I agree with the second sentence of the above quote.

>Right now, there is no reason to use super because the functions are literally placeholders. If they're suddenly not placeholders, then users that didn't use super are SOL and can't get the benefit of whatever changes you made.

this inverse of this (i.e. what you're proposing) is to force users to use the changes, by not making it an opt-in (user bumps the version which means `_beforeHook` now has `doSomethingElseBeforeDoSomething()` in it. instead of being able to `super` or not, the choice they have regarding `doSomethingElseBeforeDoSomething` is "use it or downgrade").

>If they're suddenly not placeholders, then users that didn't use super are SOL and can't get the benefit of whatever changes you made.

You're looking at this backwards. they're not "SOL", it's just not going to change behavior without them actively choosing for it to do so (by adding `super`). That's _the definition_ of backwards compatibility.

>There are issues with class inheritance from a design standpoint; not every user is going to want that.

That's true of every design choice. Are you claiming that the other proposed solution is not making any trade-offs? I reject that premise. Are you saying that your proposal serves more people better? I would like to see _any_ supporting evidence. 

>I don't see the point in shoehorning them into class inheritance or mucking with the base class to get the functionality you want to give them.

Here's an alternative to class extension that works with the current implementation (note: at no point have I ever espoused "mucking with the base class"):

```js
const stip = new Stipulate();

stip.beforeRequest = (url, options) => {
  // do whatever you want
  return [modifiedUrl, modifiedOptions];
};
```

If you're dead set on being able to pass a function as an option on instantiation, we could include the following in the constructor:
```
class Stipulate {
   constructor(options) {
    if(options.before) {
      this.beforeRequest = options.before;
    }
    if(options.after) {
      this.afterResponse = options.after;
    }
    ...
```